pkg/controller/chaosscheduler/createEngineForRepeat.go

```diff
@@ -128,50 +135,44 @@ func (schedulerReconcile *reconcileScheduler) createNewEngine(cs *chaosTypes.Sch
 	}
 	cs.Instance.Status.Schedule.StartTime = startTime
 
-	if errUpdate := schedulerReconcile.r.client.Update(context.TODO(), cs.Instance); errUpdate != nil {
-		return reconcile.Result{}, errUpdate
+	if err := schedulerReconcile.r.client.Update(context.TODO(), cs.Instance); err != nil {
+		return reconcile.Result{}, err
 	}
 
 	return reconcile.Result{}, nil
 }
 
-func getRecentUnmetScheduleTime(cs *chaosTypes.SchedulerInfo, cronString string) (*time.Time, error) {
+func (schedulerReconcile *reconcileScheduler) getRecentUnmetScheduleTime(cs *chaosTypes.SchedulerInfo, cronString string) (*time.Time, time.Duration, error) {
 
 	now := time.Now()
 	cronSchedule, err := cron.ParseStandard(cronString)
 	if err != nil {
-		return nil, fmt.Errorf("unparseable schedule: %s : %s", cronString, err)
+		return nil, 0, fmt.Errorf("unparseable schedule: %s : %s", cronString, err)
 	}
 
 	timeRange := cs.Instance.Spec.Schedule.Repeat.TimeRange
-
+	var nextTime *time.Time
 	var earliestTime time.Time
+
 	if cs.Instance.Status.LastScheduleTime != nil {
+		// handles all the schedules except first schedule
 		earliestTime = cs.Instance.Status.LastScheduleTime.Time
-	} else if timeRange != nil && timeRange.StartTime != nil {
-		// If none found, then this is either a recently created schedule,
-		// or the active/completed info was somehow lost (it may need to be recreated),
-		// or that we have started a engine, but have not noticed it yet (distributed
-		//	systems can have arbitrary delays).
-		// Since we are checking the startTime with currentTime. It is
-		// possible that we might miss a schedule at the TIME of CREATION
-		// by just a few seconds because of the slight delay in reconciliation process.
-		earliestTime = timeRange.StartTime.Time.Add(time.Minute * -1)
+	} else if timeRange != nil && timeRange.StartTime != nil && !cs.Instance.GetCreationTimestamp().Time.After(timeRange.StartTime.Time) {
+		// handles the first schedule
+		earliestTime = timeRange.StartTime.Time
 	} else {
+		// handles the first schedule
 		earliestTime = cs.Instance.GetCreationTimestamp().Time
 	}
-
-	if earliestTime.After(now) {
-		return nil, nil
+	t := cronSchedule.Next(earliestTime)
+	if !t.After(now) {
+		nextTime = &now
 	}
-	var previousTime *time.Time
-
-	for t := cronSchedule.Next(earliestTime); !t.After(now); t = cronSchedule.Next(t) {
-		temp := t
-		previousTime = &temp
+	wait := t.Sub(now)
+	if wait < 0 {
+		wait = 0
 	}
-	// cs.Instance.Status.Schedule.ExpectedNextRunTime = metav1.Time{Time: sched.Next(*previousTime)}
-	return previousTime, nil
+	return nextTime, wait, nil
```

Comment:
Why do we need to pass 3 parameters??
`wait` will be equal to `nextTime.sub(now)` right??? So can't we just return one of them either wait duration or next time??

Reply:
Yes, but nexTime is a pointer so for the cases where the next scheduled time is expected to be in the future then nextTime will be nil so we can't predict the wait time in the `createEngineRepeat` function. That's why I am deriving it and passing it

Reply:
What will be the value of `wait` if nextTime is nil??

Reply:
it will be waiting time before expected next schedule